common/src/chunk_ray_casting.rs

```diff
@@ -0,0 +1,316 @@
+use crate::{
+    collision_math::Ray,
+    math,
+    node::{ChunkLayout, Coords, VoxelAABB, VoxelData},
+    world::Material,
+};
+
+pub struct ChunkCastHit {
+    /// The tanh of the distance traveled along the ray to result in this hit.
+    pub tanh_distance: f32,
+
+    /// The coordinates of the block that was hit, including margins.
+    pub voxel_coords: Coords,
+
+    /// Which of the three axes is orthogonal to the face of the block that was hit.
+    pub face_axis: u32,
+
+    // Either +1 or -1, depending on whether the outside of the face that was hit was in the positive or
+    // negative direction in `face_axis`.
+    pub face_direction: i8,
```

Comment:
I'm ambivalent about the use of enums for that. The number inside has an actual numerical meaning (for `face_axis`, it's the array index, and for `face_direction`, it's a number you can add to the coordinate).

That being said, it does seem worth wrapping these values in tuple structs for extra type safety, perhaps combining them into a single `CubeFace` struct.

Reply:
C-style enums can have meaningful numerical values:

```rs
enum Direction {
    Plus = 1,
    Minus = -1,
}
```

You can then convert to the number with e.g `x as i8` when needed for computation.

Reply:
That's a good point. As far as I know, there isn't a nice way to convert an integer to an enum, but that's definitely fine for `Direction`, and it should also be fine for `Axis`, since I can just iterate over the values like we do for `Side` and `Vertex`.

Reply:
> As far as I know, there isn't a nice way to convert an integer to an enum

The obvious `match` expression will get erased by LLVM, and is a small price to pay for clarity, but indeed no need to bother adding the from-integer path without a use case.

> I can just iterate over the values like we do for Side and Vertex.

Yeah; IMO this is clearer anyway. Fewer magic numbers.

Reply:
Hmm, I might want to save the use of an `Axis` enum for `FaceAxis` for a future PR. The changes are more far-reaching than I originally expected. For instance, such axes would also need to be used as components in nalgebra vectors if we want to use them consistently.

The arguments I use for using a type like that could be made upstream to nalgebra, but I don't think nalgebra has a type like that. To limit scope, I would be in favor of avoiding this extra indirection layer for now.

Directions might still be worthwhile, though.

Reply:
Why not convert to a number before constructing coordinates derived from these?

Reply:
I've now added an enum change for `CoordDirection`. I'll leave this thread unresolved until we reach a decision on `CoordAxis`.

Reply:
I think I've changed my mind here. A `CoordAxis` type may be worthwhile after all. In math-heavy code that interacts with nalgebra, I prefer working with an integer type (especially since we have to deal with the `w` coordinate, which complicates the abstraction further), but most of the code in this PR (besides the ray tracer) isn't very math heavy and could benefit from a more type-safe interface.

I might give this another go, but for the sake of incremental improvement, my preference would be to only use it in places that are paired up with `CoordDirection`, such as the ray tracer return value and some of the methods in `node.rs`, but not, for instance, in `VoxelAABB`, as that seems like it should be its own PR.